### ISSUE TITLE
Mixed-domain reductions from vectors into scalars seem broken for the nonblocking backend

### ISSUE THREAD

**Issue by anyzelman** [Algebraic-Programming/ALP#305]
This is the error if reducing a vector of doubles into a size_t using a `Monoid< operators::add< double >, identities::zero >`. It compiles for all other backends, as indeed would be expected as we do NOT pass a no_casting descriptor (and even if we did, ALP should raise a clear error message instead of something like the below). The error goes away in the following cases:

- compile it with another backend;
- reduce a vector of doubles into a double scalar;
- use a mixed-data-type monoid (using `operators::add< size_t, double, size_t`  in the case of a `foldl`)

```
In file included from /home/yzelman/Documents/graphblas/include/graphblas/blas1.hpp:35,
                 from /home/yzelman/Documents/graphblas/include/graphblas.hpp:511,
                 from /home/yzelman/Documents/graphblas/tests/smoke/hook/small_simple_pagerank.cpp:18:
/home/yzelman/Documents/graphblas/include/graphblas/nonblocking/blas1.hpp: In instantiation of ‘grb::RC grb::internal::fold_from_vector_to_scalar_generic(IOType&, const grb::Vector<MaskType, grb::nonblocking, Coords>&, const grb::Vector<MaskType, grb::nonblocking, Coords>&, const Monoid&) [with unsigned int descr = 16; bool masked = false; bool left = true; Monoid = grb::Monoid<grb::operators::add<double>, grb::identities::zero>; IOType = long unsigned int; InputType = double; MaskType = bool; Coords = grb::internal::Coordinates<grb::nonblocking>]’:
/home/yzelman/Documents/graphblas/include/graphblas/nonblocking/blas1.hpp:11245:5:   required from ‘grb::RC grb::foldl(IOType&, const grb::Vector<InputType, grb::nonblocking, Coords>&, const Monoid&, const typename std::enable_if<(((! grb::is_object<InputType1>::value) && (! grb::is_object<InputType2>::value)) && grb::is_monoid<Monoid>::value), void>::type*) [with unsigned int descr = 16; Monoid = grb::Monoid<grb::operators::add<double>, grb::identities::zero>; IOType = long unsigned int; InputType = double; Coords = grb::internal::Coordinates<grb::nonblocking>; typename std::enable_if<(((! grb::is_object<InputType1>::value) && (! grb::is_object<InputType2>::value)) && grb::is_monoid<Monoid>::value), void>::type = void]’
/home/yzelman/Documents/graphblas/include/graphblas/algorithms/simple_pagerank.hpp:254:33:   required from ‘grb::RC grb::algorithms::simple_pagerank(grb::Vector<IOType>&, const grb::Matrix<NonzeroT>&, grb::Vector<IOType>&, grb::Vector<IOType>&, grb::Vector<IOType>&, grb::Vector<bool>&, IOType, IOType, size_t, size_t*, double*) [with unsigned int descr = 0; IOType = double; NonzeroT = void; size_t = long unsigned int]’
/home/yzelman/Documents/graphblas/tests/smoke/hook/small_simple_pagerank.cpp:108:56:   required from here
/home/yzelman/Documents/graphblas/include/graphblas/nonblocking/blas1.hpp:640:53: error: no matching function for call to ‘fold_from_vector_to_scalar_dense<true>(long unsigned int&, const size_t&, const size_t&, const grb::Vector<double>&, const grb::Monoid<grb::operators::add<double>, grb::identities::zero>&)’
  640 |       ret = fold_from_vector_to_scalar_dense< left >(
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
  641 |        local_reduced[ thread_id ], lower_bound, upper_bound, to_fold, monoid );
      |        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/yzelman/Documents/graphblas/include/graphblas/blas1.hpp:29,
                 from /home/yzelman/Documents/graphblas/include/graphblas.hpp:511,
                 from /home/yzelman/Documents/graphblas/tests/smoke/hook/small_simple_pagerank.cpp:18:
/home/yzelman/Documents/graphblas/include/graphblas/reference/blas1.hpp:105:6: note: candidate: ‘template<bool left, class Monoid, class InputType, class Coords> grb::RC grb::internal::fold_from_vector_to_scalar_dense(typename Monoid::D3&, const grb::Vector<InputType, grb::reference, Coords>&, const Monoid&)’
  105 |   RC fold_from_vector_to_scalar_dense(
      |      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/yzelman/Documents/graphblas/include/graphblas/reference/blas1.hpp:105:6: note:   template argument deduction/substitution failed:
In file included from /home/yzelman/Documents/graphblas/include/graphblas/blas1.hpp:35,
                 from /home/yzelman/Documents/graphblas/include/graphblas.hpp:511,
                 from /home/yzelman/Documents/graphblas/tests/smoke/hook/small_simple_pagerank.cpp:18:
/home/yzelman/Documents/graphblas/include/graphblas/nonblocking/blas1.hpp:640:53: note:   mismatched types ‘const grb::Vector<InputType, grb::reference, Coords>’ and ‘const size_t’ {aka ‘const long unsigned int’}
  640 |       ret = fold_from_vector_to_scalar_dense< left >(
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
  641 |        local_reduced[ thread_id ], lower_bound, upper_bound, to_fold, monoid );
      |        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/yzelman/Documents/graphblas/include/graphblas/reference/blas1.hpp:10082,
                 from /home/yzelman/Documents/graphblas/include/graphblas/blas1.hpp:29,
                 from /home/yzelman/Documents/graphblas/include/graphblas.hpp:511,
                 from /home/yzelman/Documents/graphblas/tests/smoke/hook/small_simple_pagerank.cpp:18:
/home/yzelman/Documents/graphblas/include/graphblas/reference/blas1.hpp:105:6: note: candidate: ‘template<bool left, class Monoid, class InputType, class Coords> grb::RC grb::internal::fold_from_vector_to_scalar_dense(typename Monoid::D3&, const grb::Vector<InputType, grb::reference_omp, Coords>&, const Monoid&)’
  105 |   RC fold_from_vector_to_scalar_dense(
      |      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/yzelman/Documents/graphblas/include/graphblas/reference/blas1.hpp:105:6: note:   template argument deduction/substitution failed:
In file included from /home/yzelman/Documents/graphblas/include/graphblas/blas1.hpp:35,
                 from /home/yzelman/Documents/graphblas/include/graphblas.hpp:511,
                 from /home/yzelman/Documents/graphblas/tests/smoke/hook/small_simple_pagerank.cpp:18:
/home/yzelman/Documents/graphblas/include/graphblas/nonblocking/blas1.hpp:640:53: note:   mismatched types ‘const grb::Vector<InputType, grb::reference_omp, Coords>’ and ‘const size_t’ {aka ‘const long unsigned int’}
  640 |       ret = fold_from_vector_to_scalar_dense< left >(
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
  641 |        local_reduced[ thread_id ], lower_bound, upper_bound, to_fold, monoid );
      |        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/yzelman/Documents/graphblas/include/graphblas/nonblocking/blas1.hpp:122:6: note: candidate: ‘grb::RC grb::internal::fold_from_vector_to_scalar_dense(typename Monoid::D3&, size_t, size_t, const grb::Vector<InputType, grb::nonblocking, Coords>&, const Monoid&) [with bool left = true; Monoid = grb::Monoid<grb::operators::add<double>, grb::identities::zero>; InputType = double; Coords = grb::internal::Coordinates<grb::nonblocking>; typename Monoid::D3 = double; size_t = long unsigned int]’ <near match>
  122 |   RC fold_from_vector_to_scalar_dense(
      |      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/yzelman/Documents/graphblas/include/graphblas/nonblocking/blas1.hpp:122:6: note:   conversion of argument 1 would be ill-formed:
/home/yzelman/Documents/graphblas/include/graphblas/nonblocking/blas1.hpp:641:21: error: cannot bind non-const lvalue reference of type ‘grb::Monoid<grb::operators::add<double>, grb::identities::zero>::D3&’ {aka ‘double&’} to an rvalue of type ‘grb::Monoid<grb::operators::add<double>, grb::identities::zero>::D3’ {aka ‘double’}
  641 |        local_reduced[ thread_id ], lower_bound, upper_bound, to_fold, monoid );
      |        ~~~~~~~~~~~~~^
/home/yzelman/Documents/graphblas/include/graphblas/nonblocking/blas1.hpp:656:22: error: cannot bind non-const lvalue reference of type ‘grb::Monoid<grb::operators::add<double>, grb::identities::zero>::D3&’ {aka ‘double&’} to an rvalue of type ‘grb::Monoid<grb::operators::add<double>, grb::identities::zero>::D3’ {aka ‘double’}
  656 |         local_reduced[ thread_id ], lower_bound, upper_bound,
      |         ~~~~~~~~~~~~~^
In file included from /home/yzelman/Documents/graphblas/include/graphblas/nonblocking/blas1.hpp:53,
                 from /home/yzelman/Documents/graphblas/include/graphblas/blas1.hpp:35,
                 from /home/yzelman/Documents/graphblas/include/graphblas.hpp:511,
                 from /home/yzelman/Documents/graphblas/tests/smoke/hook/small_simple_pagerank.cpp:18:
/home/yzelman/Documents/graphblas/include/graphblas/nonblocking/boolean_dispatcher_blas1.hpp:236:25: note:   initializing argument 3 of ‘grb::RC grb::internal::boolean_dispatcher_fold_from_vector_to_scalar_fullLoopSparse(bool, bool, typename Monoid::D3&, size_t, size_t, const Coords&, const Coords&, const grb::Vector<OutputType, grb::nonblocking, Coords>&, const grb::Vector<InputType, grb::nonblocking, Coords>&, const Monoid&) [with unsigned int descr = 16; bool masked = true; bool left = true; Monoid = grb::Monoid<grb::operators::add<double>, grb::identities::zero>; InputType = double; MaskType = bool; Coords = grb::internal::Coordinates<grb::nonblocking>; typename Monoid::D3 = double; size_t = long unsigned int]’
  236 |    typename Monoid::D3 &thread_local_output,
      |    ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
In file included from /home/yzelman/Documents/graphblas/include/graphblas/blas1.hpp:35,
                 from /home/yzelman/Documents/graphblas/include/graphblas.hpp:511,
                 from /home/yzelman/Documents/graphblas/tests/smoke/hook/small_simple_pagerank.cpp:18:
/home/yzelman/Documents/graphblas/include/graphblas/nonblocking/blas1.hpp:687:23: error: cannot bind non-const lvalue reference of type ‘grb::Monoid<grb::operators::add<double>, grb::identities::zero>::D3&’ {aka ‘double&’} to an rvalue of type ‘grb::Monoid<grb::operators::add<double>, grb::identities::zero>::D3’ {aka ‘double’}
  687 |          local_reduced[ thread_id ], lower_bound, upper_bound,
      |          ~~~~~~~~~~~~~^
In file included from /home/yzelman/Documents/graphblas/include/graphblas/nonblocking/blas1.hpp:53,
                 from /home/yzelman/Documents/graphblas/include/graphblas/blas1.hpp:35,
                 from /home/yzelman/Documents/graphblas/include/graphblas.hpp:511,
                 from /home/yzelman/Documents/graphblas/tests/smoke/hook/small_simple_pagerank.cpp:18:
/home/yzelman/Documents/graphblas/include/graphblas/nonblocking/boolean_dispatcher_blas1.hpp:82:25: note:   initializing argument 3 of ‘grb::RC grb::internal::boolean_dispatcher_fold_from_vector_to_scalar_vectorDriven(bool, bool, typename Monoid::D3&, size_t, size_t, const Coords&, const Coords&, const grb::Vector<OutputType, grb::nonblocking, Coords>&, const grb::Vector<InputType, grb::nonblocking, Coords>&, const Monoid&) [with unsigned int descr = 16; bool masked = false; bool left = true; Monoid = grb::Monoid<grb::operators::add<double>, grb::identities::zero>; InputType = double; MaskType = bool; Coords = grb::internal::Coordinates<grb::nonblocking>; typename Monoid::D3 = double; size_t = long unsigned int]’
   82 |    typename Monoid::D3 &thread_local_output,
      |    ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
In file included from /home/yzelman/Documents/graphblas/include/graphblas/blas1.hpp:35,
                 from /home/yzelman/Documents/graphblas/include/graphblas.hpp:511,
                 from /home/yzelman/Documents/graphblas/tests/smoke/hook/small_simple_pagerank.cpp:18:
/home/yzelman/Documents/graphblas/include/graphblas/nonblocking/blas1.hpp:703:23: error: cannot bind non-const lvalue reference of type ‘grb::Monoid<grb::operators::add<double>, grb::identities::zero>::D3&’ {aka ‘double&’} to an rvalue of type ‘grb::Monoid<grb::operators::add<double>, grb::identities::zero>::D3’ {aka ‘double’}
  703 |          local_reduced[ thread_id ], lower_bound, upper_bound,
      |          ~~~~~~~~~~~~~^
In file included from /home/yzelman/Documents/graphblas/include/graphblas/nonblocking/blas1.hpp:53,
                 from /home/yzelman/Documents/graphblas/include/graphblas/blas1.hpp:35,
                 from /home/yzelman/Documents/graphblas/include/graphblas.hpp:511,
                 from /home/yzelman/Documents/graphblas/tests/smoke/hook/small_simple_pagerank.cpp:18:
/home/yzelman/Documents/graphblas/include/graphblas/nonblocking/boolean_dispatcher_blas1.hpp:236:25: note:   initializing argument 3 of ‘grb::RC grb::internal::boolean_dispatcher_fold_from_vector_to_scalar_fullLoopSparse(bool, bool, typename Monoid::D3&, size_t, size_t, const Coords&, const Coords&, const grb::Vector<OutputType, grb::nonblocking, Coords>&, const grb::Vector<InputType, grb::nonblocking, Coords>&, const Monoid&) [with unsigned int descr = 16; bool masked = false; bool left = true; Monoid = grb::Monoid<grb::operators::add<double>, grb::identities::zero>; InputType = double; MaskType = bool; Coords = grb::internal::Coordinates<grb::nonblocking>; typename Monoid::D3 = double; size_t = long unsigned int]’
  236 |    typename Monoid::D3 &thread_local_output,
      |    ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
In file included from /home/yzelman/Documents/graphblas/include/graphblas/blas1.hpp:35,
                 from /home/yzelman/Documents/graphblas/include/graphblas.hpp:511,
                 from /home/yzelman/Documents/graphblas/tests/smoke/hook/small_simple_pagerank.cpp:18:
/home/yzelman/Documents/graphblas/include/graphblas/nonblocking/blas1.hpp:721:23: error: cannot bind non-const lvalue reference of type ‘grb::Monoid<grb::operators::add<double>, grb::identities::zero>::D3&’ {aka ‘double&’} to an rvalue of type ‘grb::Monoid<grb::operators::add<double>, grb::identities::zero>::D3’ {aka ‘double’}
  721 |          local_reduced[ thread_id ], lower_bound, upper_bound,
      |          ~~~~~~~~~~~~~^
In file included from /home/yzelman/Documents/graphblas/include/graphblas/nonblocking/blas1.hpp:53,
                 from /home/yzelman/Documents/graphblas/include/graphblas/blas1.hpp:35,
                 from /home/yzelman/Documents/graphblas/include/graphblas.hpp:511,
                 from /home/yzelman/Documents/graphblas/tests/smoke/hook/small_simple_pagerank.cpp:18:
/home/yzelman/Documents/graphblas/include/graphblas/nonblocking/boolean_dispatcher_blas1.hpp:158:25: note:   initializing argument 3 of ‘grb::RC grb::internal::boolean_dispatcher_fold_from_vector_to_scalar_maskDriven(bool, bool, typename Monoid::D3&, size_t, size_t, const Coords&, const Coords&, const grb::Vector<InputType, grb::nonblocking, Coords>&, const grb::Vector<MaskType, grb::nonblocking, Coords>&, const Monoid&) [with unsigned int descr = 16; bool left = true; Monoid = grb::Monoid<grb::operators::add<double>, grb::identities::zero>; InputType = double; MaskType = bool; Coords = grb::internal::Coordinates<grb::nonblocking>; typename Monoid::D3 = double; size_t = long unsigned int]’
  158 |    typename Monoid::D3 &thread_local_output,
      |    ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/9/functional:59,
                 from /home/yzelman/Documents/graphblas/include/graphblas/reference/vector.hpp:27,
                 from /home/yzelman/Documents/graphblas/include/graphblas/vector.hpp:32,
                 from /home/yzelman/Documents/graphblas/include/graphblas.hpp:500,
                 from /home/yzelman/Documents/graphblas/tests/smoke/hook/small_simple_pagerank.cpp:18:
/usr/include/c++/9/bits/std_function.h:667:7: error: ‘std::function<_Res(_ArgTypes ...)>::function(_Functor) [with _Functor = grb::internal::fold_from_vector_to_scalar_generic(IOType&, const grb::Vector<MaskType, grb::nonblocking, Coords>&, const grb::Vector<MaskType, grb::nonblocking, Coords>&, const Monoid&) [with unsigned int descr = 16; bool masked = false; bool left = true; Monoid = grb::Monoid<grb::operators::add<double>, grb::identities::zero>; IOType = long unsigned int; InputType = double; MaskType = bool; Coords = grb::internal::Coordinates<grb::nonblocking>]::<lambda(grb::internal::Pipeline&, size_t, size_t)>; <template-parameter-2-2> = void; <template-parameter-2-3> = void; _Res = grb::RC; _ArgTypes = {grb::internal::Pipeline&, long unsigned int, long unsigned int}]’, declared using local type ‘grb::internal::fold_from_vector_to_scalar_generic(IOType&, const grb::Vector<MaskType, grb::nonblocking, Coords>&, const grb::Vector<MaskType, grb::nonblocking, Coords>&, const Monoid&) [with unsigned int descr = 16; bool masked = false; bool left = true; Monoid = grb::Monoid<grb::operators::add<double>, grb::identities::zero>; IOType = long unsigned int; InputType = double; MaskType = bool; Coords = grb::internal::Coordinates<grb::nonblocking>]::<lambda(grb::internal::Pipeline&, size_t, size_t)>’, is used but never defined [-fpermissive]
  667 |       function<_Res(_ArgTypes...)>::
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/yzelman/Documents/graphblas/include/graphblas/reference/vector.hpp:42,
                 from /home/yzelman/Documents/graphblas/include/graphblas/vector.hpp:32,
                 from /home/yzelman/Documents/graphblas/include/graphblas.hpp:500,
                 from /home/yzelman/Documents/graphblas/tests/smoke/hook/small_simple_pagerank.cpp:18:
/home/yzelman/Documents/graphblas/include/graphblas/blas0.hpp:391:12: warning: ‘grb::RC grb::foldl(IOType&, const InputType&, const OP&, const typename std::enable_if<((grb::is_operator<OP>::value && (! grb::is_object<InputType1>::value)) && (! grb::is_object<InputType2>::value)), void>::type*) [with unsigned int descr = 16; OP = grb::operators::add<double>; InputType = double; IOType = long unsigned int]’ used but never defined
  391 |  static RC foldl(
      |            ^~~~~
/home/yzelman/Documents/graphblas/include/graphblas/blas0.hpp:391:12: warning: ‘grb::RC grb::foldl(IOType&, const InputType&, const OP&, const typename std::enable_if<((grb::is_operator<OP>::value && (! grb::is_object<InputType1>::value)) && (! grb::is_object<InputType2>::value)), void>::type*) [with unsigned int descr = 16; OP = grb::operators::add<double>; InputType = long unsigned int; IOType = double]’ used but never defined
make[3]: *** [tests/smoke/CMakeFiles/test_small_pagerank_nonblocking.dir/build.make:96: tests/smoke/CMakeFiles/test_small_pagerank_nonblocking.dir/hook/small_simple_pagerank.cpp.o] Error 1
```

**Comment by anyzelman** [comment]
Assigning to @aristeidis-mastoras -- seems like the one most suited to handle this one;) If not ok time-wise though we can discuss who else could take it over~

**Comment by anyzelman** [comment]
Crossref: @aristeidis-mastoras pushed a fix to #261

**Comment by anyzelman** [comment]
Fix merged with #261.
